Background process support

RyanGibb opened the pull request:
Extends the wallpaper setter to support long lived processes as described in https://github.com/bcyran/timewall/issues/140.

We 'make before break', creating a new process to set the background before killing the old. I've created a cleanup function that's used by the preview command to remove the last process setting the wallpaper. This could also be used by the daemon if signal handlers were implemented (though it seems a few dependencies would be required to do this). Notably the non-daemon set command does not clean up after itself.

I've tested this for a short lived command and it all seems to work okay. The only limitation is we aren't capturing it's stderr output anymore. I think we could create a thread to wait on the process to finish and print it's error status if that's an issue.

Do let me know if I'm holding Rust wrong anywhere, or if there's stylistic things to fix.

bcyran commented:
Thanks for the PR!

As for the implementation: I have to admit I'm a bit bothered by the lack of cleanup, especially by the fact that one-shot `timewall set` will just spawn new processes with every invocation. Maybe we could store the PID in some temp file and use it to kill the previous process on subsequent `timewall set` runs? We could also add a command like `timewall unset` to kill the currently running process.

As for the stylistic stuff: lints and formatting checks fail. I enabled workflows for PRs on master so I think it should be visible in CI for the PR if you rebase? Not sure if it will work, I don't have much experience with GH CI stuff. You can also run all the checks locally with `just` command. `just format` will autoformat the code.

RyanGibb commented:
> As for the implementation: I have to admit I'm a bit bothered by the lack of cleanup, especially by the fact that one-shot `timewall set` will just spawn new processes with every invocation. Maybe we could store the PID in some temp file and use it to kill the previous process on subsequent `timewall set` runs? We could also add a command like `timewall unset` to kill the currently running process.

Agreed, I've pushed an update fixing this.

> As for the stylistic stuff: lints and formatting checks fail. I enabled workflows for PRs on master so I think it should be visible in CI for the PR if you rebase? Not sure if it will work, I don't have much experience with GH CI stuff. You can also run all the checks locally with `just` command. `just format` will autoformat the code.

The formatter appears to pass locally now!

I've also added a grace period to wait one second after the new process had been created before killing the old, so that the new one has time to take effect. Maybe this period should be configurable?

RyanGibb commented:
> I've also added a grace period to wait one second after the new process had been created before killing the old, so that the new one has time to take effect. This period could be made configurable.

Actually, this proved problematic for previews with short delays, since the cache would get updated before the forked thread had killed the PID. Instead I've moved the delay inline in the set wallpaper command function.

I've also added the delay parameter to the set command so the daemon can also use this behaviour.

bcyran commented:
Nice, I tried this out and I think we're on a good track with this.

Is that okay with you if I take it over from here? There's some stuff here I don't have any experience with and I'd like to play around with different options. Also, I don't want to annoy you with my nitpicks :wink:.

RyanGibb commented:
Go ahead! I've tidied up the PR (removing the mutable self to WallpaperSetter) and squashed the commit history.
